discovery_server.go

```diff
@@ -0,0 +1,314 @@
+	Start() error
+
+	// List returns the list of the currently available ports. It works
+	// only after a Start.
+	List() (portList []*Port, err error)
+
+	// StartSync is called to put the discovery in event mode. When the
+	// function returns the discovery must send port events ("add" or "remove")
+	// using the eventCB function.
+	StartSync(eventCB EventCallback) error
+
+	// Stop stops the discovery internal subroutines. If the discovery is
+	// in event mode it must stop sending events through the eventCB previously
+	// set.
+	Stop() error
+
+	// Quit is called just before the server terminates. This function can be
+	// used by the discovery as a last chance gracefully close resources.
+	Quit()
+}
+
+// EventCallback is a callback function to call to transmit port
+// metadata when the discovery is in "sync" mode and a new event
+// is detected.
+type EventCallback func(event string, port *Port)
+
+// A DiscoveryServer is a pluggable discovery protocol handler,
+// it must be created using the NewDiscoveryServer function.
+type DiscoveryServer struct {
+	impl               Discovery
+	out                io.Writer
+	outMutex           sync.Mutex
+	userAgent          string
+	reqProtocolVersion int
+	initialized        bool
+	started            bool
+	syncStarted        bool
+	syncChannel        chan interface{}
+}
+
+// NewDiscoveryServer creates a new discovery server backed by the
+// provided pluggable discovery implementation. To start the server
+// use the Run method.
+func NewDiscoveryServer(impl Discovery) *DiscoveryServer {
+	return &DiscoveryServer{
+		impl: impl,
+	}
+}
+
+// Run starts the protocol handling loop on the given input and
+// output stream, usually `os.Stdin` and `os.Stdout` are used.
+// The function blocks until the `QUIT` command is received or
+// the input stream is closed. In case of IO error the error is
+// returned.
+func (d *DiscoveryServer) Run(in io.Reader, out io.Writer) error {
+	d.out = out
+	reader := bufio.NewReader(in)
+	for {
+		fullCmd, err := reader.ReadString('\n')
+		if err != nil {
+			d.outputError("command_error", err.Error())
+			return err
+		}
+		split := strings.Split(fullCmd, " ")
+		cmd := strings.ToUpper(strings.TrimSpace(split[0]))
+
+		if !d.initialized && cmd != "HELLO" {
+			d.outputError("command_error", fmt.Sprintf("First command must be HELLO, but got '%s'", cmd))
+			continue
+		}
+
+		switch cmd {
+		case "HELLO":
+			d.hello(fullCmd[6:])
+		case "START":
+			d.start()
+		case "LIST":
+			d.list()
+		case "START_SYNC":
+			d.startSync()
+		case "STOP":
+			d.stop()
+		case "QUIT":
+			d.impl.Quit()
+			d.outputOk("quit")
+			return nil
+		default:
+			d.outputError("command_error", fmt.Sprintf("Command %s not supported", cmd))
+		}
+	}
+}
+
+func (d *DiscoveryServer) hello(cmd string) {
+	if d.initialized {
+		d.outputError("hello", "HELLO already called")
+		return
+	}
+	re := regexp.MustCompile(`(\d+) "([^"]+)"`)
```

Comment:
Shouldn't we change the regexp to `^(\d+) "([^"]+)$`?  Otherwise it could match strings like `BEFORE 1 "user agent" AFTER`.